From 7b4fff406064f8c91dee69ca4c4dde9e60c8532a Mon Sep 17 00:00:00 2001
From: Pankaj Dev <pankaj.dev@st.com>
Date: Wed, 27 Sep 2023 15:37:34 +0530
Subject: [PATCH 364/954] fix(st-usb): optimize dwc3 structure size

Locate bounce_buf outside endpoint struct, since for operation
its only needed for OUT_EPs, also the bounce_buf needs to be
allocated on cache-line aligned size, hence locating it as first
variable in dwc3_handle struct could help reduce padding

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git f118577e2b957ec4702d985d2c5225b9137e4123 fix(st-usb): optimize dwc3 structure size]
Signed-off-by: Pankaj Dev <pankaj.dev@st.com>
Change-Id: I0eff6d206df9db72cc8910ab11db95bdccf672fa
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/331131
Domain-Review: Fabrice GASNIER <fabrice.gasnier@foss.st.com>
Reviewed-by: Fabrice GASNIER <fabrice.gasnier@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Tested-by: Yann GAUTIER <yann.gautier@foss.st.com>
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
---
 drivers/st/usb_dwc3/usb_dwc3.c |  2 ++
 include/drivers/st/usb_dwc3.h  | 19 ++++++++-----------
 2 files changed, 10 insertions(+), 11 deletions(-)

diff --git a/drivers/st/usb_dwc3/usb_dwc3.c b/drivers/st/usb_dwc3/usb_dwc3.c
index 020566db9c..03735658eb 100644
--- a/drivers/st/usb_dwc3/usb_dwc3.c
+++ b/drivers/st/usb_dwc3/usb_dwc3.c
@@ -2343,6 +2343,8 @@ void usb_dwc3_init_driver(struct usb_handle *usb_core_handle, struct pcd_handle
 		dwc3_handle->OUT_ep[i].phy_epnum = __HAL_PCD_EPADDR_TO_PHYEPNUM(i);
 
 		dwc3_handle->OUT_ep[i].intr_num = PCD_DEV_EVENTS_INTR;
+
+		dwc3_handle->OUT_ep[i].bounce_buf = dwc3_handle->bounce_bufs[i].bounce_buf;
 	}
 
 #define PHYS_AREA	STM32MP_USB_DWC3_BASE
diff --git a/include/drivers/st/usb_dwc3.h b/include/drivers/st/usb_dwc3.h
index 824c497fb1..47bdc76592 100644
--- a/include/drivers/st/usb_dwc3.h
+++ b/include/drivers/st/usb_dwc3.h
@@ -169,16 +169,8 @@ typedef struct {
 	uint8_t flags;            /* Endpoint state flags condition */
 	uint8_t phy_epnum;        /* physical EP number 0 - 31 */
 	uint8_t intr_num;         /* Interrupt number to get events for this EP */
-#ifdef CACHED_MEMORY // Needed only for cached (__KERNEL__) environment
-    /* Added to shield the cache line (8 words) intersection with other members of structure */
-	uint32_t dummy3[CACHE_LINE_SIZE];
-#endif
 	/* to be used for receiving packet < EP-Max-Packet-Size */
-	uint8_t bounce_buf[USB3_MAX_PACKET_SIZE];
-#ifdef CACHED_MEMORY // Needed only for cached (__KERNEL__) environment
-	/* Added to shield the cache line (8 words) intersection with other members of structure */
-	uint32_t dummy4[CACHE_LINE_SIZE];
-#endif
+	uint8_t *bounce_buf;
 	/* Transmission FIFO number. Number between Min_Data = 1 and Max_Data = 15 */
 	uint8_t tx_fifo_num;
 	uint8_t resc_idx;       /* Resource index */
@@ -189,6 +181,10 @@ typedef struct {
 	uint32_t xfer_dest_len;
 } usb_dwc3_endpoint_t;
 
+typedef struct {
+	uint8_t bounce_buf[USB3_MAX_PACKET_SIZE];  /* Event Buffer area */
+} usb_dwc3_endpoint_bouncebuf_t __aligned(CACHE_LINE_SIZE);
+
 /* Aligned to make EvtBuffer start address evtbuffer-size aligned */
 typedef struct {
 	uint8_t EvtBufferArea[USB_DWC3_EVENT_BUFFER_SIZE];  /* Event Buffer area */
@@ -219,9 +215,10 @@ typedef enum {
 
 typedef struct {
 	/*
-	 * Kept at top to avoid struct size wastage since EvtBufferArea
-	 * needs to be USB_DWC3_EVENT_BUFFER_SIZE aligned
+	 * Kept at top to avoid struct size wastage since bounce_buf
+	 * needs to be CACHE_LINE_SIZE aligned
 	 */
+	usb_dwc3_endpoint_bouncebuf_t bounce_bufs[USB_DWC3_NUM_OUT_EP];
 	PCD_intbuffersdef intbuffers;
 	usb_dwc3_global_t *usb_global;
 	usb_dwc3_device_t *usb_device;
-- 
2.53.0

